src/photon.cpp

```diff
@@ -162,24 +167,35 @@ PhotonInteraction::PhotonInteraction(hid_t group)
       close_dataset(dset);
 
       int n_transition = dims[0];
-      shell.n_transitions = n_transition;
       if (n_transition > 0) {
         xt::xtensor<double, 2> matrix;
         read_dataset(tgroup, "transitions", matrix);
 
-        shell.transition_subshells =
-          xt::view(matrix, xt::all(), xt::range(0, 2));
-        shell.transition_energy = xt::view(matrix, xt::all(), 2);
-        shell.transition_probability = xt::view(matrix, xt::all(), 3);
-        shell.transition_probability /= xt::sum(shell.transition_probability)();
+        // Transition probability normalization
+        double norm = xt::sum(xt::col(matrix, 3))();
+
+        shell.transitions.resize(n_transition);
+        for (int j = 0; j < n_transition; ++j) {
+          auto& transition = shell.transitions[j];
+          transition.primary_subshell = shell_map.at(matrix(j, 0));
+          transition.secondary_subshell = shell_map.at(matrix(j, 1));
+          transition.energy = matrix(j, 2);
+          transition.probability = matrix(j, 3) / norm;
+        }
       }
-    } else {
-      shell.n_transitions = 0;
     }
     close_group(tgroup);
   }
   close_group(rgroup);
 
+  // Check the maximum size of the atomic relaxation stack
+  auto max_size = this->calc_max_stack_size();
+  if (max_size > MAX_STACK_SIZE && mpi::master) {
+    warning("The subshell vacancy stack in atomic relaxation can grow up to " +
+            std::to_string(max_size) + ", but the stack size limit is set to " +
+            std::to_string(MAX_STACK_SIZE) + ".");
```

Comment:
You can use `fmt::format` here
```suggestion
    warning(fmt::format(
      "The subshell vacancy stack in atomic relaxation can grow up to {}, but "
      "the stack size limit is set to {}.",
      max_size, MAX_STACK_SIZE));
```
You'll need `#include <fmt/core.h>` up top